### PR TITLE
Improve breadcrumb bar accessibility

### PR DESCRIPTION
**Pull request by janfaracik**
**What's new** 
* The active breadcrumb is now a text element rather than a clickable link (better for accessibility)
* Cleanup to the code to bring it inline with modern Jenkins BEM practices 
* ARIA labels have been added

---

**Behaviour changes**
* The active breadcrumb (due to it being text and not a link) no longer has a dropdown menu
  * I'd argue the menu wasn't necessary as it just duplicates what's on the page
  * For the Configure System page, the section-to-sidebar-items solution would be better suited
* The chevron after the active breadcrumb has been removed if the active crumb is not a ModelObjectWithChildren

### Proposed changelog entries

*  N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6912"></a>

### CODE REVIEW

**Comment by uhafner** [comment]
> **Behaviour changes**
> 
> * The active breadcrumb (due to it being text and not a link) no longer has a dropdown menu
>   
>   * I'd argue the menu wasn't necessary as it just duplicates what's on the page
>   * For the Configure System page, the section-to-sidebar-items solution would be better suited
> * The chevron after the active breadcrumb has been removed
>   
>   * It suggests there is more after the active breadcrumb, but that's never the case
> 
> Would be good to hear thoughts on the above ^

I think that both improvements make sense and clean up the UI!

**Comment by NotMyFault** [comment]
> **Behaviour changes**
> * The active breadcrumb (due to it being text and not a link) no longer has a dropdown menu
>   * I'd argue the menu wasn't necessary as it just duplicates what's on the page
>   * For the Configure System page, the section-to-sidebar-items solution would be better suited

Makes sense to remove it, it displays content you already have on the page available.

> * The chevron after the active breadcrumb has been removed 
>   * It suggests there is more after the active breadcrumb, but that's never the case 

+1

File: war/src/main/less/abstracts/theme.less
Comment (NotMyFault) [inline war/src/main/less/abstracts/theme.less:74]: The removed variables have two usages in the dark theme and in the solarized theme plugin.

Objections about removing or adjusting those in the plugins?

Comment (janfaracik) [reply war/src/main/less/abstracts/theme.less:74]: I've updated the dark-theme plugin https://github.com/jenkinsci/dark-theme-plugin/pull/219

Will take a look at Solarized now..

Comment (NotMyFault) [reply war/src/main/less/abstracts/theme.less:74]: Thanks 👍🏻

Comment (janfaracik) [reply war/src/main/less/abstracts/theme.less:74]: Solarized update is here: https://github.com/jenkinsci/solarized-theme-plugin/pull/60

**Review by NotMyFault — APPROVED** [review]

**Comment by jtnord** [comment]
To me the screenshot looks like the breadcrumbs are now disabled (non interactive) which removes the visual hint that you can interact with them, which to me is not a good thing.

> The active breadcrumb (due to it being text and not a link) no longer has a dropdown menu
I'd argue the menu wasn't necessary as it just duplicates what's on the page

but the breadcrumb was a much more concise and easier to use way to get to it - with less scrolling (esp on long pages where want you want is somewhere in the middle between several graphs).


> The chevron after the active breadcrumb has been removed

but there can be things after it like: 

image

I'm sure there is an API to do this and it is not a one off hack.

**Comment by janfaracik** [comment]
> To me the screenshot looks like the breadcrumbs are now disabled (non interactive) which removes the visual hint that you can interact with them, which to me is not a good thing.

I understand your point, we need some way to visually separate the breadcrumbs so I went with dimming the inactive ones. I'd worry that if we were to make them look like links (e.g. blue) we'd draw too much attention to them, and if we were to highlight the active breadcrumb it might look interactive which it isn't. It's a tricky one.

> but the breadcrumb was a much more concise and easier to use way to get to it - with less scrolling (esp on long pages where want you want is somewhere in the middle between several graphs).

I understand that too, I think that the breadcrumb menus are more of a bandaid to the problem however, the problem being overall page complexity/poor use of space. 

Using a sidebar like the new config screen/plugin manager is a better use of the real estate imo and it's also more obvious for the user, rather than a dropdown menu in the breadcrumb - a pattern which I think is exclusive to Jenkins. The project/build screens need a lot of work (if not a total rework) and I think it'd be great to see it reorganised so it's faster for the user to get to what they need.

**Comment by jtnord** [comment]
> Using a sidebar like the new config screen/plugin manager is a better use of the real estate imo and it's also more obvious for the user

that would not work very well (if at all) for Multibranch projects or folders.  I would not expect it's children to be listed there.

FWIW windows explorer does show the chevron after the last item to allow you to selevt children so it is not an unexpected paradigm

image


image

**Comment by basil** [comment]
The gray background intended for the default theme applies even when I choose the Dark Theme, which makes Jenkins unusable in the Dark Theme.

**Comment by janfaracik** [comment]
> that would not work very well (if at all) for Multibranch projects or folders. I would not expect it's children to be listed there.

> FWIW windows explorer does show the chevron after the last item to allow you to select children so it is not a completely unexpected / unusual paradigm

Ah, I had forgotten Windows had that - haven't properly used a Windows machine for a few years now. 

I've restored the last chevron (if it's clickable), we can always take another look at it in the future once some layout improvements roll in to Jenkins. Hope that addresses your issue :) 

---

> The gray background intended for the default theme applies even when I choose the Dark Theme, which makes Jenkins unusable in the Dark Theme.

There's a PR open for that https://github.com/jenkinsci/dark-theme-plugin/pull/219 :)

**Comment by basil** [comment]
Thanks Jan, sorry I missed that - I searched this page for "dark" but didn't search the dark theme plugin's open pull requests.

**Comment by timja** [comment]
> I've restored the last chevron (if it's clickable), we can always take another look at it in the future once some layout improvements roll in to Jenkins. Hope that addresses your issue :)

Doesn't seem to be working right.

e.g. take a freestyle job and click the last chevron (which is there now) it does nothing.

image


whereas on weekly.ci.jenkins.io it does:
image

**Review by timja — DISMISSED** [review]
LGTM.

ATH is passing and manual testing all looks good to me.

**Comment by timja** [comment]
@NotMyFault can you re-check please?

**Comment by timja** [comment]
Actually I found an issue.

the context menu isn't working for credentials on:
http://localhost:8080/manage/credentials/

e.g. 
There's no context menu for:
image

on weekly.ci.jenkins.io:

image

**Review by timja — APPROVED** [review]
I see that's called out as a behaviour change:
> The active breadcrumb (due to it being text and not a link) no longer has a dropdown menu
> I'd argue the menu wasn't necessary as it just duplicates what's on the page

I guess probably okay, weirdly it's not created on the page but credentials plugin really needs a refresh to use UI components appropriately

**Comment by timja** [comment]
(FWIW I'm working on a credentials plugin improvement currently in this area)

**Comment by NotMyFault** [comment]
Minor nit: The default view uses the wrong color.

Screenshot 2022-09-04 at 20 41 11

**Comment by timja** [comment]
(credentials plugin PR that makes removing the breadcrumb from there okay: https://github.com/jenkinsci/credentials-plugin/pull/364)

**Review by NotMyFault — APPROVED** [review]
Awesome job 🎉

**Comment by jtnord** [comment]
So a quick run through this locally and it all seems to work, however I still believe that the grey for breadcrumbs is not the right choice.

As you can not actually click on the current breadcrumb (it would just take you to the page you are on) so in essence we highlight the current breadcrumb using the same colour as we do for interactive links (in the sidebar) , but it is the only non interactive breadcrumb out of all of them.   Given the "disabled" appearance of the other breadcrumbs I feel this is a backward step, leading to a worse discoverability and adding to cognitive overload (it is disabled if x or y but not in the breadcrumb.  its a link in most of jenkins but not if it is text in the main panel or in the breadcrumb bar)

For example - https://getbootstrap.com/docs/4.0/components/breadcrumb/  shows that all the bread crumbs except the current page crumb should be highlighted, which is a more usual paradigm esp given you can interact with everything *except* the current element.  I will note however that https://www.adobe.com/uk/products/photoshop.html uses the opposite of that (which is what this PR uses).


from the description
> The chevron after the active breadcrumb has been removed
It suggests there is more after the active breadcrumb, but that's never the case

I think this is now out dated - if the active crumb is a `ModelObjectWithChildren` the crumb is still shown.

**Comment by cristina-pizzagalli** [comment]
FWIW, I agree with @jtnord 's point about the inversion of typical link colours. A user will generally expect the active links to be the darker colour, with disabled links being lighter. If the colours used could be swapped, I think it would help improve the usability of the breadcrumb nav. Thanks :)

**Comment by janfaracik** [comment]
> So a quick run through this locally and it all seems to work, however I still believe that the grey for breadcrumbs is not the right choice.
> 
> As you can not actually click on the current breadcrumb (it would just take you to the page you are on) so in essence we highlight the current breadcrumb using the same colour as we do for interactive links (in the sidebar) , but it is the only non interactive breadcrumb out of all of them. Given the "disabled" appearance of the other breadcrumbs I feel this is a backward step, leading to a worse discoverability and adding to cognitive overload (it is disabled if x or y but not in the breadcrumb. its a link in most of jenkins but not if it is text in the main panel or in the breadcrumb bar)

I 100% get where you're coming from, my only gripe is that for me that colour scheme just doesn't feel right for the breadcrumb bar. Maybe that's just a me thing though, like you mentioned though there are a lot of different design systems that handle it in various different ways - there isn't one ideal way of approaching it.

Screenshot 2022-09-27 at 00 12 52

vs

Screenshot 2022-09-27 at 00 15 52

How would you feel about me undoing the change in colour for now and coming back to it at a later date? It's only a couple of colour vars so it wouldn't be too much pain to come back to (especially when more of Jenkins gets restyled etc). That way we could get the improved accessibility features in.

> I think this is now out dated - if the active crumb is a `ModelObjectWithChildren` the crumb is still shown.

I've updated the PR description, and I've also added the aria labels from the Bootstrap link you sent - thanks! :)

**Comment by timja** [comment]
Out of the two screenshots you posted the second one looks better to me

**Comment by jtnord** [comment]
> How would you feel about me undoing the change in colour for now and coming back to it at a later date?

probably better so this can move on.

Out of the 2 screenshots I prefer the first - however would also expect the `>` to be dark as they are also interactable elements.
That said I wonder if things could be improved with a more radical change than just colour, but as we said for a discussion elsewhere but leaving here just to get people thinking (and I expect several people to say OMG that is ugly!).

image

**Comment by janfaracik** [comment]
> > How would you feel about me undoing the change in colour for now and coming back to it at a later date?
> 
> probably better so this can move on.

Awesome, I've removed the styling changes so the breadcrumb items look the same as they did before now.

> I think the main issue we have is that the breadcrumb bar is not just a breadcrumb bar. It has navigation from the element (to the page), from the end of the element (to children) and for some things from and extra bit that is neither the element nor the end of the element (the context menu).

Agreed. There was an attempt at simplifying the breadcrumb bar quite a while back now in #6084 in which the two dropdown menus were merged into one - would be good to take another look at that and menus in general in Jenkins in the future.

**Comment by timja** [comment]
@jtnord are you able to re-review this?

**Comment by timja** [comment]
> I won't be able to look at this today, however based on

doesn't have to be today, did you want to look before merge or should we ship and iterate if needed?

**Review by jtnord — APPROVED** [review]
span up locally and looks good to me.

I had to disable `yarn lint` as I needed to merge in master (without it many plugins failed due to the `TriggerItem` api introduced in core.  not sure exactly what failed - the eslint file was too large for vscodes little head but it was not the usual git gave me windows line ends (it did but I had fixed those manually)!
(in other words may be worth a synch with master to ensure the CI is still good before merging)

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by benipeled** [comment]
As mentioned above by @jtnord, the side-effect of this change is that the last object in the bar is not clickable at - ex. the build number - and this is kind of "regression" since it was quite useful in case of long pages (console) that we want to quickly return to the top of the page,

@janfaracik can we add back the clickable behavior?

**Comment by MarkEWaite** [comment]
> As mentioned above by @jtnord, the side-effect of this change is that the last object in the bar is not clickable at - ex. the build number - and this is kind of "regression" since it was quite useful in case of long pages (console) that we want to quickly return to the top of the page,
> 
> @janfaracik can we add back the clickable behavior?

It is actually not a side effect, but one of the key items in the pull request.  @janfaracik said in the bullets describing the pull request that the last object in the bar is intentionally not clickable.

> The active breadcrumb is now a text element rather than a clickable link (better for accessibility)

It shows the user that it is the current page by not allowing click on the element.  The accessibility decision was further discussed in the UX SIG meeting and agreed there to remain with the final element in breadcrumb trail not being clickable.

**Comment by uhafner** [comment]
>  it was quite useful in case of long pages (console) that we want to quickly return to the top of the page,
> 

I think you revealed a bug in the Console view, can you please file that as Jira issue? Normally, each Jenkins view should create its own breadcrumb link, it seems that this link is missing for the console view (it does not create a new breadcrumb sub item for the current view). 

Broken console view:

Bildschirm­foto 2022-11-29 um 09 42 44

Correctly implemented view example:

Bildschirm­foto 2022-11-29 um 09 43 23

Additionally, you can create a feature request in Jenkins Jira that describes your (in my opinion) actual requirement: From your description it looks like you basically want to have a quick navigation in the console log where you can simply jump to the fist and last line of the log? A good option would be some navigation buttons that allow to jump to the first and last line of the log.

**Comment by amuniz** [comment]
Definitely a bug in the console log view. There is no way to navigate back to the build page (other than adjusting the URL in the browser).

**Comment by lemeurherve** [comment]
For reference, same kind of issue as the one resolved in https://github.com/jenkinsci/cloudbees-folder-plugin/pull/278

**Comment by Wadeck** [comment]
=> https://issues.jenkins.io/browse/JENKINS-70169 for the console log view

**Comment by daniel-beck** [comment]
> The active breadcrumb is now a text element rather than a clickable link (better for accessibility)

This is going to be a problem for any non-`index` view on a `ModelObject` unless they explicitly declare a breadcrumb.

* I cannot get from `/manage/computer/configure` to `/manage/computer/`
* I cannot get from `/view/all/builds` to `/view/all/`

There's probably others. WDYT about only doing that when on the `index` view? If you're not on the `index` view and it's the last breadcrumb, it's not a no-op after all.

**Comment by uhafner** [comment]
> > The active breadcrumb is now a text element rather than a clickable link (better for accessibility)
> 
> This is going to be a problem for any non-`index` view on a `ModelObject` unless they explicitly declare a breadcrumb.
> 

Maybe that is the actual problem: why do we have these special views? Shouldn't all views get a breadcrumb link by default?

**Comment by daniel-beck** [comment]
> Shouldn't all views get a breadcrumb link by default?

How would you infer the label?

**Comment by daniel-beck** [comment]
> `getDisplayName` of `ModelObject`

That'll just duplicate the previous breadcrumb that always exists. Please read the cases I describe above again. For example, in https://weekly.ci.jenkins.io/view/all/builds, you're asking to have another breadcrumb labeled _All_.

**Comment by lemeurherve** [comment]
I'm currently walking through views needing a breadcrumb while preparing my PR to fix most (all?) of them, there isn't an easy way to set its correct title unfortunately, it needs a case per case in many places.

**Comment by uhafner** [comment]
> > `getDisplayName` of `ModelObject`
> 
> That'll just duplicate the previous breadcrumb that always exists. Please read the cases I describe above again. For example, in https://weekly.ci.jenkins.io/view/all/builds, you're asking to have another breadcrumb labeled _All_.

I'm not sure if I understand your argumentation. The view that renders the URL `builds` should be a `ModelObject` with the name `Build History of Jenkins`. Or is this a view without a model object?

**Comment by daniel-beck** [comment]
> Or is this a view without a model object?

Correct. The `ModelObject` is the `View`. Otherwise there would never be a use case for a non-`index` view.

**Comment by uhafner** [comment]
I see. So the same `ModelObject` renders different URLs? This is something that we should avoid. Has this approach a historical background or some use cases that could not be done by providing a standalone model object for each link? In my plugins every view is a model object with URL and display name. I must admit that I don't see the benefit for not following a strict architectural pattern that every view with a URL should be a model object. Would it be feasible to change that or at least create a warning if someone does not follow the default path?

**Comment by uhafner** [comment]
> > This is something that we should avoid.
> 
> It has always worked like this and any attempt to change this will require changing practically every plugin that defines custom views. Just to make sure this is clear: _Every_ view not named `index` that is an actual view (~has `<l:layout>`) is affected by this. I wouldn't be surprised if it's thousands across the plugin ecosystem. Often it doesn't make sense to update breadcrumbs. For example, `PluginManager` is the `ModelObject` for the `index` (updates, really), `available`, `installed`, and `advanced` views. Having each view governed by its own `ModelObject` would be madness.

Tanks for clarifying. In my opinion having a `ModelObject` for each view would be a clean architecture and not madness. From #7457 we see that we now need to change a lot of views, this anti-pattern clearly violates the open closed principle. But anyway, it's too late for an architectural change for the existing views. I can only recommend to implement it differently for new views.

**Comment by Wadeck** [comment]
⚠️ Due to the attempt to easily fix the issues in core in https://github.com/jenkinsci/jenkins/pull/7457, I am wondering if we should revert this PR and provide a more organized migration path. I already have seen multiple feedbacks about users being surprised by the new behavior, which is not necessarily a good signal.

**Comment by janfaracik** [comment]
Hey folks, 

Thought I'd comment on a few points just to share my thoughts:

1. Giving every page a breadcrumb item

I'd be opposed to giving every page their own breadcrumb item, IMO pages inside of the two column navigation sidebar (like the build page, plugin manager, design library) should only have the one breadcrumb (that being the parent page). I see the sidebar for switching part of a pages content, much like a tab bar, hence not having an additional breadcrumb for its children pages.

To go back from the console output, clicking 'Status' should work just fine (unless there's a plugin/situation where that's not possible?)

----

2. Usage of the breadcrumb button as a way to jump to the top of the page

In a similar fashion to the use of the breadcrumb bar item to refresh the page (sorry again!) the component wasn't designed to jump to the top of the page, although I can definitely see how that'd be helpful in both cases. In the current prototype design for Jenkins there's a button to do so in the sidebar which might be of interest to you.

image

As for a possible reversion - if this is wanted (and I have no quarrel with that if that's preferred) I'd rather myself revert the change to the last breadcrumb (from a link to text) rather than revert the entire PR as there's still benefits to the component outside of the mentioned change.

**Comment by benipeled** [comment]
> In the current prototype design for Jenkins there's a button to do so in the sidebar which might be of interest to you.

That might be an nice option,

Another option is to stick the sidebar on scrolling down (position: fixed), allowing easy navigation between the different components (console, config etc.)

**Comment by daniel-beck** [comment]
> From https://github.com/jenkinsci/jenkins/pull/7457 we see that we now need to change a lot of views

We can just revert part of this PR.

**Comment by jtnord** [comment]
> > Shouldn't all views get a breadcrumb link by default?
> 
> How would you infer the label?

if the URL is not the URL of the last breadcrumb - use the page title as a fallback if not specified?  

Will be a little "verbose" in many cases "Manage Jenkins -> Plugin Manager -> Plugin Manage Updates"  (or something like that) 

(possibly will break somewhere though, and there is the question of if trying to infer a breadcrumb is even a good idea or not)

**Comment by jtnord** [comment]
(OT: to go to the top of the page with the mouse there is a handy "reload" button in almost every browser just above the breadcrumb bar) for some accessibility there is keyboard for this too (ctrl+r, and F5 on windows (option something on mac?)) for assistive tech, not sure but I would expect a "reload" option to controll the browser too

**Comment by daniel-beck** [comment]
> use the page title as a fallback if not specified

In case of https://weekly.ci.jenkins.io/view/all/builds, that would still result in a duplicate "All" link. Now what?

> without reloading there is ctrl+home (on windows - something with option on mac)

Just Home, which is Fn+Left on small keyboards. Also for iOS it's tapping near the top of the screen.

Given the myriad options this feels like less of a necessity to me, although I guess some pages with dynamically loaded content still benefit from easy reloading. But that seems to me that the pages themselves should be looked at critically.

> no standard mouse option to do it in one.

Firefox has reload in the context menu.

**Comment by benipeled** [comment]
> Another option is to stick the sidebar on scrolling down (position: fixed), allowing easy navigation between the different components (console, config etc.)

@janfaracik multiple options on the sidebar should be taken into account but in general this is the idea
WDYT?

[fixed_sidebar.webm](https://user-images.githubusercontent.com/31824530/204810177-a8dad7bc-8500-4237-a122-a4e349afaf2d.webm)

**Comment by uhafner** [comment]
> > Another option is to stick the sidebar on scrolling down (position: fixed), allowing easy navigation between the different components (console, config etc.)
> 
> @janfaracik multiple options on the sidebar should be taken into account but in general this is the idea WDYT?
> 

Yes, this is the concept we should use for the sidebar.

**Comment by janfaracik** [comment]
> Another option is to stick the sidebar on scrolling down (position: fixed)

I'm definitely in favour of the sidebar being stuck to the side of the screen 👍 The project config screen already does this, would be good to get it adopted by other pages consistently (only issue is pages with sidebars that are too long).

**Comment by timja** [comment]
see https://groups.google.com/g/jenkinsci-dev/c/2Z1OuI6XC90/m/Se_zVkfwGwAJ for a mailing list request of a revert or partial revert, @janfaracik are you able to take a look at this please?

**Pull request merged**
